xarray/core/weighted.py

```diff
@@ -0,0 +1,239 @@
+
+    Parameters
+    ----------
+    dim : str or sequence of str, optional
+        Dimension(s) over which to apply the weighted `{fcn}`.
+    skipna : bool, optional
+        If True, skip missing values (as marked by NaN). By default, only
+        skips missing values for float dtypes; other dtypes either do not
+        have a sentinel missing value (int) or skipna=True has not been
+        implemented (object, datetime64 or timedelta64).
+    keep_attrs : bool, optional
+        If True, the attributes (`attrs`) will be copied from the original
+        object to the new one.  If False (default), the new object will be
+        returned without attributes.
+
+    Returns
+    -------
+    reduced : {cls}
+        New {cls} object with weighted `{fcn}` applied to its data and
+        the indicated dimension(s) removed.
+    """
+
+_SUM_OF_WEIGHTS_DOCSTRING = """
+    Calculate the sum of weights, accounting for missing values in the data
+
+    Parameters
+    ----------
+    dim : str or sequence of str, optional
+        Dimension(s) over which to sum the weights.
+    keep_attrs : bool, optional
+        If True, the attributes (`attrs`) will be copied from the original
+        object to the new one.  If False (default), the new object will be
+        returned without attributes.
+
+    Returns
+    -------
+    reduced : {cls}
+        New {cls} object with the sum of the weights over the given dimension.
+    """
+
+
+class Weighted:
+    """A object that implements weighted operations.
+
+    You should create a Weighted object by using the `DataArray.weighted` or
+    `Dataset.weighted` methods.
+
+    See Also
+    --------
+    Dataset.weighted
+    DataArray.weighted
+    """
+
+    __slots__ = ("obj", "weights")
+
+    @overload
+    def __init__(self, obj: "DataArray", weights: "DataArray") -> None:
+        ...
+
+    @overload  # noqa: F811
+    def __init__(self, obj: "Dataset", weights: "DataArray") -> None:  # noqa: F811
+        ...
+
+    def __init__(self, obj, weights):  # noqa: F811
+        """
+        Create a Weighted object
+
+        Parameters
+        ----------
+        obj : DataArray or Dataset
+            Object over which the weighted reduction operation is applied.
+        weights : DataArray
+            An array of weights associated with the values in the obj.
+            Each value in the obj contributes to the reduction operation
+            according to its associated weight.
+
+        Note
+        ----
+        ``weights`` must be a ``DataArray`` and cannot contain missing values.
+        Missing values can be replaced by `weights.fillna(0)`.
+        """
+
+        from .dataarray import DataArray
+
+        assert isinstance(weights, DataArray), "'weights' must be a DataArray"
+
+        self.obj = obj
+
+        if weights.isnull().any():
+            raise ValueError("`weights` cannot contain missing values.")
+
+        self.weights = weights
```

Comment:
we should check that `weights.sizes[dim] == obj.sizes[dim]` for all weighted dims.

Reply:
Do you know if there is a `is_broadcastable` function?

Reply:
Not that I'm aware of. https://stackoverflow.com/questions/47243451/checking-if-two-arrays-are-broadcastable-in-python

In this case the simple check should be enough. something like `all(weights.sizes[dim] == obj.sizes[dim] for dim in weights if dim in obj)`

Reply:
Do you think `da` and `weights` must align? Also #3694